### PR TITLE
[YARN] Add yarn.lock support

### LINKED ISSUE #238: [YARN] Yarn support?
A lot of folks use yarn.  It looks like your current strategy of reading the lockfile will work fine, it just needs to be smart about whether it is reading a package-lock or a yarn.lock file.  Most of the code should work fine though!

### PR DESCRIPTION
**Pull request by koconnor-dev**
Dynamically convert the _yarn.lock_ file to a _package-lock.json_ format using _synp_, so as to support yarn... should fix #238

### CODE REVIEW

File: tests/integration/with-yarn-lockfile/package.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  },
+  "dependencies": {
+    "packageurl-js": "^0.0.1",
+    "parse-packagejson-name": "^1.0.1",
+    "prettify-xml": "^1.2.0",
+    "read-installed": "^4.0.3",
+    "ssri": "^6.0.0",
+    "uuid": "^3.3.2",
+    "xmlbuilder": "^13.0.2",
+    "xmldom": "^0.2.1"
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:53]: @me ignore

File: tests/integration/with-yarn-lockfile/yarn.lock [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+  version "1.0.7"
+  resolved "https://registry.npmjs.org/path-parse/-/path-parse-1.0.7.tgz#fbc114b60ca42b30d9daf5858e4bd68bbedb6735"
+  integrity sha512-LDJzPVEEEPR+y48z93A0Ed0yXb8pAByGWo/k5YYdYgpY2/2EsOsksJrq7lOHxryrVOn1ejG6oAp8ahvOIQD8sw==
+
+prettify-xml@^1.2.0:
+  version "1.2.0"
+  resolved "https://registry.npmjs.org/prettify-xml/-/prettify-xml-1.2.0.tgz#46dcf1ee8a8d8b73db30b7e06ef26dc9cf3f6f18"
+  integrity sha1-Rtzx7oqNi3PbMLfgbvJtyc8/bxg=
+
+read-installed@^4.0.3:
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:171]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+  version "1.0.2"
+  resolved "https://registry.npmjs.org/wrappy/-/wrappy-1.0.2.tgz#b5243d8f3ec1aa35f1364605bc0d1036e30ab69f"
+  integrity sha1-tSQ9jz7BqjXxNkYFvA0QNuMKtp8=
+
+xmlbuilder@^13.0.2:
+  version "13.0.2"
+  resolved "https://registry.npmjs.org/xmlbuilder/-/xmlbuilder-13.0.2.tgz#02ae33614b6a047d1c32b5389c1fdacb2bce47a7"
+  integrity sha512-Eux0i2QdDYKbdbA6AM6xE4m6ZTZr4G4xF9kahI2ukSEMCzwce2eX9WlTI5J3s+NU7hpasFsr8hWIONae7LluAQ==
+
+xmldom@^0.2.1:
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:295]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:53]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    "url": "git+https://github.com/CycloneDX/cyclonedx-node-module.git"
+  },
+  "bugs": {
+    "url": "https://github.com/CycloneDX/cyclonedx-node-module/issues"
+  },
+  "dependencies": {
+    "packageurl-js": "^0.0.1",
+    "parse-packagejson-name": "^1.0.1",
+    "prettify-xml": "^1.2.0",
+    "read-installed": "^4.0.3",
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:49]: @sonatype-lift ignore

**Comment by jkowalleck** [comment]
Hello @koconnor-dev ,

thank you for the feature and this pull request.

Could you sign your commits, according to the [CONTRIBUTING](https://github.com/CycloneDX/cyclonedx-node-module/blob/master/CONTRIBUTING.md) file?
Hint: https://github.com/CycloneDX/cyclonedx-node-module/pull/282/checks?check_run_id=5987696158

**Comment by jkowalleck** [comment]
TODO:

- [ ] have yarn added to the CI /CT engine
- [ ] have the test setup constants renamed, to ake clear which wone is yarn and which one is npm. 
- [ ] have all tests running nad passing in CI/CT

File: tests/integration/setup.js [hunk 5]
```diff
@@ -28,6 +28,10 @@ const REQUIRES_INSTALL = [
   'with-packages'
 ]
 
+const REQUIRES_YARN_INSTALL = [
```
Comment (jkowalleck) [inline tests/integration/setup.js:31]: please rename the other const `REQUIRES_INSTALL` to `REQUIRES_NPM_INSTALL`

File: tests/integration/setup.js [hunk 6]
```diff
@@ -45,3 +49,19 @@ for (const DIR of REQUIRES_INSTALL) {
     console.error(done)
   }
 }
+
+let yarnDone
```
Comment (jkowalleck) [inline tests/integration/setup.js:53]: use the already existing `done` variable.

**Comment by koconnor-dev** [comment]
> Hello @koconnor-dev ,
> 
> thank you for the feature and this pull request.
> 
> Could you sign your commits, according to the [CONTRIBUTING](https://github.com/CycloneDX/cyclonedx-node-module/blob/master/CONTRIBUTING.md) file? Hint: https://github.com/CycloneDX/cyclonedx-node-module/pull/282/checks?check_run_id=5987696158

I'll try 😃

File: tests/integration/with-yarn-lockfile/package.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:50]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 3 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:295]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 2 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:171]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    "ssri": "^6.0.0",
+    "uuid": "^3.3.2",
+    "xmlbuilder": "^13.0.2",
+    "xmldom": "^0.2.1"
```
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:54]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:50]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 3 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:295]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 2 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:171]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:54]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:50]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 3 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:295]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/yarn.lock (hunk 2 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/yarn.lock:171]: @sonatype-lift ignore

File: tests/integration/with-yarn-lockfile/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (koconnor-dev) [reply tests/integration/with-yarn-lockfile/package.json:54]: @sonatype-lift ignore

**Comment by jkowalleck** [comment]
feature was released via https://github.com/CycloneDX/cyclonedx-node-module/releases/tag/v3.7.0

**Comment by koconnor-dev** [comment]
Please note that the yarn conversion is done by "https://www.npmjs.com/package/synp", therefore whatever yarn versions it supports apply here... this is a late response to issues raised #238 and #294

**Comment by jkowalleck** [comment]
Tanks, @koconnor-dev.

current state of yarn support in this project is locked.
Future implementations and improvements should be done here: https://github.com/CycloneDX/cyclonedx-node-yarn

**Pull request merged**